### PR TITLE
feat: add user management code

### PR DESCRIPTION
**Pull request by ClearloveHn**

### CODE REVIEW

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -8,13 +8,13 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v2
-      - name: Set up JDK 21
+      - name: Set up JDK 8
```
Comment (cr7258) [inline .github/workflows/codeql-analysis.yml]:
```suggestion
      - name: Set up JDK 21
```

File: .github/workflows/codeql-analysis.yml [hunk 2]
```diff
@@ -8,13 +8,13 @@ jobs:
...
         uses: actions/setup-java@v1
         with:
-          java-version: 21
+          java-version: 8
```
Comment (cr7258) [inline .github/workflows/codeql-analysis.yml]:
```suggestion
          java-version: 21
```

File: build.gradle [hunk 3]
```diff
@@ -1,25 +1,38 @@
...
 }
 
 group = 'pro.quicksense'
 version = '0.0.1-SNAPSHOT'
 
 java {
-    sourceCompatibility = '21'
+    sourceCompatibility = '1.8'
+    targetCompatibility = '1.8'
 }
```
Comment (cr7258) [inline build.gradle]: Can we use relatively new Java and Spring versions?

File: src/main/java/pro/quicksense/modules/common/CommonConstant.java [hunk 4]
```diff
@@ -0,0 +1,580 @@
+package pro.quicksense.modules.common;
+
+/**
+ * @Description: 通用常量
+ * @author: jeecg-boot
+ */
+public interface CommonConstant {
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/common/CommonConstant.java]: Please remove unused constant.

File: src/main/java/pro/quicksense/modules/common/Result.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+/**
+ *   接口返回数据格式
+ * @author scott
+ * @email jeecgos@163.com
+ * @date  2019年1月19日
+ */
+@Data
+@ApiModel(value="接口返回对象", description="接口返回对象")
+public class Result<T> implements Serializable {
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/common/Result.java]: Please remove unused variables and functions.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+import java.util.UUID;
+
+/**
+ * <p>
+ * 用户管理
+ * </p>
+ *
+ * @Author lhn
+ * @since 2024-06-09
+ */
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Remove unnecessary comments.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    @Autowired
+    private ISysUserService sysUserService;
+
+
+    /**
+     * 用户注册
+     *
+     * @param jsonObject
+     * @return
+     */
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Remove unnecessary comments.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+    @PostMapping("/register")
+    public Result<JSONObject> userRegister(@RequestBody JSONObject jsonObject) {
+        Result<JSONObject> result = new Result<JSONObject>();
+        String phone = jsonObject.getString("phone");
+        String username = jsonObject.getString("username");
+        String email = jsonObject.getString("email");
+
+        // 未设置用户名，则用手机号作为用户名
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Add comments in English.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+        }
+
+        // 未设置密码，则随机生成一个密码
+        String password = jsonObject.getString("password");
+        if (StringUtils.isBlank(password)) {
+            password = RandomUtil.randomString(8);
+        }
+
+        // 校验用户名、手机号、邮箱是否已注册
+        SysUser sysUser1 = sysUserService.getOne(Wrappers.<SysUser>query().lambda().eq(SysUser::getUsername, username));
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Use descriptive variable names, e.g., `existingUsername`

File: src/main/java/pro/quicksense/modules/eneity/SysUser.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+ * <p>
+ * 用户表
+ * </p>
+ *
+ * @Author scott
+ * @since 2018-12-20
+ */
+@Data
+@TableName("sys_user")
+public class SysUser implements Serializable {
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/eneity/SysUser.java]: How about using `User`?

File: src/main/java/pro/quicksense/modules/common/Result.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+     * 兼容VUE3版token失效不跳转登录页面
+     * @param code
+     * @param message
+     */
+    public Result(Integer code, String message) {
+        this.code = code;
+        this.message = message;
+    }
+
+    public Result<T> success(String message) {
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/common/Result.java]: For all functions in use, please add the corresponding unit test.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+
+        if (StringUtils.isBlank(email)) {
+            SysUser sysUser3 = sysUserService.getOne(Wrappers.<SysUser>query().lambda().eq(SysUser::getEmail, email));
+            if (sysUser3 != null) {
+                result.setMessage("邮箱已被注册");
+                result.setSuccess(false);
+                return result;
+            }
+        }
+
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: email verification code

File: src/main/java/pro/quicksense/modules/mapper/xml/SysUserMapper.xml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<!DOCTYPE mapper
+        PUBLIC "-//mybatis.org//DTD Mapper 3.0//EN"
+        "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/mapper/xml/SysUserMapper.xml]: Remove

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 14]
```diff
@@ -1,271 +1,181 @@
...
-     * @param jsonObject
-     * @return
-     */
     @PostMapping("/register")
-    public Result<JSONObject> userRegister(@RequestBody JSONObject jsonObject) {
-        Result<JSONObject> result = new Result<JSONObject>();
-        String phone = jsonObject.getString("phone");
+    public Result<?> userRegister(@RequestBody JSONObject jsonObject) {
+        Result<?> result = new Result<>();
         String username = jsonObject.getString("username");
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: A not-null judgment is recommended.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 15]
```diff
@@ -1,271 +1,181 @@
...
             user.setUsername(username);
-            user.setRealName(realname);
-            user.setPassword(passwordEncode);
-            user.setSalt(salt);
+            user.setRealName(realName);
+            user.setPassword(encodedPassword);
             user.setEmail(email);
-            user.setPhone(phone);
-            user.setStatus(CommonConstant.USER_UNFREEZE);
+            user.setStatus(CommonConstant.USER_STATUS_FROZEN);
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Consider a constructor to simplify the code.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 16]
```diff
@@ -1,271 +1,181 @@
...
 
 
-    /**
-     * 用户注册
-     *
-     * @param jsonObject
-     * @return
-     */
     @PostMapping("/register")
-    public Result<JSONObject> userRegister(@RequestBody JSONObject jsonObject) {
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Consider moving the logic to a '@Service' class.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 17]
```diff
@@ -1,271 +1,181 @@
...
+        User user = new User();
+
+        if (StringUtils.isNotBlank(id)) {
+            user = sysUserService.getById(id);
+        } else {
+            List<User> userList = sysUserService.list();
+            Result<List<User>> result = new Result<>();
+            result.setCode(200);
+            result.setMsg("User list");
+            result.setData(userList);
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Consider a constructor to make the code more concise.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 18]
```diff
@@ -1,271 +1,181 @@
...
-        if (StringUtils.isBlank(password)) {
-            password = RandomUtil.randomString(8);
+        String confirmPassword = jsonObject.getString("confirmPassword");
+        String realName = jsonObject.getString("realName");
+
+        // Validate email address format
+        String emailPattern = "^[a-zA-Z0-9_-]+@[a-zA-Z0-9_-]+(\\.[a-zA-Z0-9_-]+)+$";
+        if (!email.matches(emailPattern)) {
+            result.setCode(400);
+            result.setMsg("Invalid email format");
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: There are Chinese msg and English msg, which should be uniform.

File: src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java [hunk 19]
```diff
@@ -1,46 +1,41 @@
...
-        if (CommonConstant.USER_FREEZE.equals(sysUser.getStatus())) {
-            result.error500("该用户已冻结");
+        //情况2：根据用户信息查询，该用户已冻结
+        if (CommonConstant.USER_STATUS_NORMAL.equals(user.getStatus())) {
+            result.setCode(500);
+            result.setMsg("该用户已冻结");
             return result;
         }
+
+        result.setCode(200);
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java]: Consider a const or enum.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 20]
```diff
@@ -1,271 +1,181 @@
...
+        BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();
+        String encodedPassword = passwordEncoder.encode(password);
 
         try {
-            SysUser user = new SysUser();
-            user.setCreateTime(new Date());// 设置创建时间
-            String salt = UUID.randomUUID().toString().replace("-", "").substring(0, 8); // 生成随机盐
-            String passwordEncode = PasswordUtil.encrypt(username, password, salt);
+            User user = new User();
+            user.setCreateTime(new Date());
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Only createTime, no updateTime?

**Review by Mchen1995 — COMMENTED** [review]
Please view the comments.

File: src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+
+
+@Service
+@Slf4j
+public class SysUserServiceImpl extends ServiceImpl<SysUserMapper, User> implements ISysUserService {
+    @Override
+    public Result<String> checkUserIsEffective(User user) {
+        Result<String> result = new Result<>();
+
+        //情况1：根据用户信息查询，该用户不存在
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java]:
```suggestion
        // The user does not exist
```

File: src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+        if (ObjectUtil.isEmpty(user)) {
+            result.setCode(500);
+            result.setMsg("该用户不存在，请注册");
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java]:
```suggestion
            result.setMsg("The user does not exist. Please sign up.");
```

File: src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+            return result;
+        }
+
+        //情况2：根据用户信息查询，该用户已冻结
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java]:
```suggestion
        // The user account is frozen.
```

File: src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+        if (CommonConstant.USER_STATUS_NORMAL.equals(user.getStatus())) {
+            result.setCode(500);
+            result.setMsg("该用户已冻结");
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java]:
```suggestion
            result.setMsg("The user account is frozen.");
```

File: src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+
+        //情况1：根据用户信息查询，该用户不存在
+        if (ObjectUtil.isEmpty(user)) {
+            result.setCode(500);
+            result.setMsg("该用户不存在，请注册");
+            return result;
+        }
+
+        //情况2：根据用户信息查询，该用户已冻结
+        if (CommonConstant.USER_STATUS_NORMAL.equals(user.getStatus())) {
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java]:
```suggestion
        if (CommonConstant.USER_STATUS_FROZEN.equals(user.getStatus())) {
```

File: src/main/resources/init.sql [hunk 26]
```diff
@@ -0,0 +1,13 @@
...
+    `real_name` VARCHAR(255),
+    `password` VARCHAR(255),
+    `salt` VARCHAR(255),
+    `email` VARCHAR(255),
+    `phone` VARCHAR(255),
+    `status` INT,
+    `create_time` DATETIME,
+    `update_time` DATETIME,
+    PRIMARY KEY (`id`)
+    ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
```
Comment (cr7258) [inline src/main/resources/init.sql]:
```suggestion
CREATE TABLE IF NOT EXISTS `user` (
    `id` VARCHAR(255) NOT NULL,
    `username` VARCHAR(255),
    `realname` VARCHAR(255),
    `password` VARCHAR(255),
    `salt` VARCHAR(255),
    `email` VARCHAR(255),
    `phone` VARCHAR(255),
    `status` INT,
    `create_time` DATETIME,
    `update_time` DATETIME,
    PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
  
```

File: src/main/java/pro/quicksense/modules/eneity/User.java [hunk 27]
```diff
@@ -0,0 +1,46 @@
+package pro.quicksense.modules.eneity;
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/eneity/User.java]: Please correct the package name. `eneity` -> `entity`

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 28]
```diff
@@ -1,271 +1,181 @@
...
+            result.setCode(400);
+            result.setMsg("Username or email already exists");
         } catch (Exception e) {
-            result.error500("注册失败");
+            result.setCode(500);
+            result.setMsg("An error occurred during registration");
         }
+
         return result;
     }
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Move this logic to Service.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 29]
```diff
@@ -28,111 +31,134 @@ public class SysUserController {
...
         String confirmPassword = jsonObject.getString("confirmPassword");
+        if (StringUtils.isBlank(confirmPassword)) {
+            return Result.error(CommonConstant.BAD_REQUEST_CODE, "Confirm password cannot be empty");
+        }
         String realName = jsonObject.getString("realName");
+        if (StringUtils.isBlank(realName)) {
+            return Result.error(CommonConstant.BAD_REQUEST_CODE, "Real name cannot be empty");
+        }
+
+        String verifyCode = jsonObject.getString("verifyCode ");
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: An extra space after "verifyCode".

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 30]
```diff
@@ -28,111 +31,134 @@ public class SysUserController {
...
 
-        Result<String> loginResult = new Result<>();
         // Login successful, generate token and return
-        userInfo(user, loginResult);
-        return loginResult;
+        return userInfo(user);
     }
 
+    @PostMapping("/loginByEmail")
+    public Result<?> logout(@RequestBody JSONObject jsonObject) {
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Can we replace the argument `JSONObject jsonobject` with a model class, like `User user` in method `edit(User user)`

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 31]
```diff
@@ -28,111 +31,134 @@ public class SysUserController {
...
+        return userInfo(user);
+
+    }
+
+
     @PostMapping("/edit")
     public Result<Object> edit(@RequestBody User user) {
-        Result<Object> result = new Result<>();
         try {
+            user.setUpdateTime(new Date());
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: This line of code could be moved into method `sysUserService.updateById(User user)`.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 32]
```diff
@@ -143,27 +169,25 @@ public Result<?> info(@RequestParam(value = "id", required = false) String id) {
...
+        return Result.success("User info", user);
     }
 
-    private Result<String> userInfo(User user, Result<String> result) {
-        String token = generateToken(user);
-
-        result.setCode(200);
-        result.setMsg("Login success");
-        result.setData(token);
+    @GetMapping("/sendEmail")
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Sending an email would change the status of the server, a `@PostMapping` is more appropriate than a `@GetMapping`.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 33]
```diff
@@ -28,111 +31,134 @@ public class SysUserController {
...
+            return Result.error(CommonConstant.BAD_REQUEST_CODE, "Password cannot be empty");
+        }
         String confirmPassword = jsonObject.getString("confirmPassword");
+        if (StringUtils.isBlank(confirmPassword)) {
+            return Result.error(CommonConstant.BAD_REQUEST_CODE, "Confirm password cannot be empty");
+        }
         String realName = jsonObject.getString("realName");
+        if (StringUtils.isBlank(realName)) {
+            return Result.error(CommonConstant.BAD_REQUEST_CODE, "Real name cannot be empty");
+        }
```
Comment (Mchen1995) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]: Similar codes, consider encapsulating them into a method.

**Review by Mchen1995 — COMMENTED** [review]
FYI

File: src/main/java/pro/quicksense/modules/service/ISysUserService.java [hunk 34]
```diff
@@ -0,0 +1,12 @@
+package pro.quicksense.modules.service;
+
+import com.baomidou.mybatisplus.extension.service.IService;
+import pro.quicksense.modules.common.Result;
+import pro.quicksense.modules.entity.User;
+
+
+public interface ISysUserService extends IService<User> {
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/service/ISysUserService.java]:
```suggestion
public interface UserService extends IService<User> {
```
also modify the file name.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 35]
```diff
@@ -0,0 +1,205 @@
...
+import java.util.Objects;
+
+
+@Slf4j
+@RestController
+@RequestMapping("/user")
+public class SysUserController {
+
+    @Autowired
+    private ISysUserService sysUserService;
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]:
```suggestion
public class UserController {

    @Autowired
    private UserService UserService;
```
also modify the file name.

File: src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java [hunk 36]
```diff
@@ -0,0 +1,69 @@
...
+import pro.quicksense.modules.mapper.UserMapper;
+import pro.quicksense.modules.service.ISysUserService;
+
+
+import java.util.Date;
+
+
+@Service
+@Slf4j
+public class SysUserServiceImpl extends ServiceImpl<UserMapper, User> implements ISysUserService {
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/service/impl/SysUserServiceImpl.java]:
```suggestion
public class UserServiceImpl extends ServiceImpl<UserMapper, User> implements UserService {
```

File: src/main/java/pro/quicksense/modules/util/SecurityConfig.java [hunk 37]
```diff
@@ -0,0 +1,22 @@
...
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        http
+                .csrf(csrf -> csrf.disable())
+                .authorizeHttpRequests(authz -> authz
+                        .requestMatchers("/user/register", "/user/login").permitAll()
+                        .anyRequest().authenticated()
+                );
+        return http.build();
+    }
+}
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/util/SecurityConfig.java:22]: I think we don't need to set CSRF here. We will have a proxy in front of the frontend and backend services to provide the same entrypoint.

File: src/main/java/pro/quicksense/modules/common/Result.java [hunk 38]
```diff
@@ -0,0 +1,57 @@
...
+        return result;
+    }
+
+    public static <T> Result<T> error(Integer code, String msg, T data) {
+        Result<T> result = new Result<>();
+        result.setCode(code);
+        result.setMsg(msg);
+        result.setData(data);
+        return result;
+    }
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/common/Result.java]:
```suggestion
```
remove unused method.

File: src/main/java/pro/quicksense/modules/common/Result.java [hunk 39]
```diff
@@ -0,0 +1,57 @@
...
+
+    public static <T> Result<T> success(T data) {
+        Result<T> result = new Result<>();
+        result.setCode(200);
+        result.setMsg("Success");
+        result.setData(data);
+        return result;
+    }
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/common/Result.java]:
```suggestion
    public static <T> Result<T> success(String msg) {
        Result<T> result = new Result<>();
        result.setCode(200);
        result.setMsg(msg);
        return result;
    }

    public static <T> Result<T> success(String msg, T data) {
        Result<T> result = new Result<>();
        result.setCode(200);
        result.setMsg(msg);
        result.setData(data);
        return result;
    }
```
Simplify success method, all success and error response should contain message.

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 40]
```diff
@@ -0,0 +1,205 @@
...
+        String confirmPassword = jsonObject.getString("confirmPassword");
+        if (StringUtils.isBlank(confirmPassword)) {
+            return Result.error(CommonConstant.BAD_REQUEST_CODE, "Confirm password cannot be empty");
+        }
+        String realName = jsonObject.getString("realName");
+        if (StringUtils.isBlank(realName)) {
+            return Result.error(CommonConstant.BAD_REQUEST_CODE, "Real name cannot be empty");
+        }
+
+        String verifyCode = jsonObject.getString("verifyCode ");
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]:
```suggestion
        String verifyCode = jsonObject.getString("verifyCode");
```

File: src/main/java/pro/quicksense/modules/common/Result.java [hunk 41]
```diff
@@ -0,0 +1,57 @@
+package pro.quicksense.modules.common;
+
+import lombok.Data;
+
+import java.io.Serializable;
+
+@Data
+public class Result<T> implements Serializable {
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/common/Result.java]: It seems like we don't need to reinvent the wheel; the Spring Boot framework provides out-of-the-box methods.

example:  
https://zetcode.com/springboot/responseentity/

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java (hunk 30 above)
Comment (cr7258) [reply src/main/java/pro/quicksense/modules/controller/SysUserController.java]: example: https://springjava.com/spring-boot/response-entity-in-rest-api-crud-example-spring-boot/

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java (hunk 31 above)
Comment (cr7258) [reply src/main/java/pro/quicksense/modules/controller/SysUserController.java]: controller:
```java
UserService.update(User user)
```

service:

```bash
update(User user) {
  user.setUpdateTime(new Date());
  UserService.updateById(user);
}
```

File: src/main/java/pro/quicksense/modules/controller/SysUserController.java [hunk 42]
```diff
@@ -0,0 +1,205 @@
...
+            user = sysUserService.getById(id);
+        } else {
+            List<User> userList = sysUserService.list();
+            return Result.success(userList);
+        }
+
+        return Result.success("User info", user);
+    }
+
+    @GetMapping("/sendEmail")
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/controller/SysUserController.java]:
```suggestion
    @PostMapping("/sendEmail")
```

File: src/main/java/pro/quicksense/modules/service/impl/UserServiceImpl.java [hunk 43]
```diff
@@ -0,0 +1,216 @@
...
+@Slf4j
+@RequiredArgsConstructor
+public class UserServiceImpl extends ServiceImpl<UserMapper, User> implements UserService {
+
+    private final EmailUtil emailUtil;
+    private final CodeUtil codeUtil;
+    private final JwtInterceptor jwtInterceptor;
+
+
+    @Transactional
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/service/impl/UserServiceImpl.java]: Do we need `@Transactional` annotation here? If `this.save(user)` fails, it means that the data has not yet been inserted into the database.

File: src/main/java/pro/quicksense/modules/service/impl/UserServiceImpl.java [hunk 44]
```diff
@@ -0,0 +1,216 @@
...
+            return ResponseEntity.status(CommonConstant.INTERNAL_SERVER_ERROR_CODE).body("Username or email already exists");
+        } catch (Exception e) {
+            log.error("Error during user registration", e);
+            return ResponseEntity.status(CommonConstant.INTERNAL_SERVER_ERROR_CODE).body("An error occurred during registration");
+        }
+
+        return ResponseEntity.ok("User registration successful");
+    }
+
+    @Override
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/service/impl/UserServiceImpl.java]: Do we need `@Override` annotation here? `login` method is not defined in the `import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl`

File: src/main/java/pro/quicksense/modules/service/impl/UserServiceImpl.java [hunk 45]
```diff
@@ -0,0 +1,216 @@
...
+        ResponseEntity<?> checkResult = checkUserIsEffective(user);
+        if (checkResult != null) {
+            return checkResult;
+        }
+
+        // Check if the username or password is correct
+        BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();
+        boolean matches = passwordEncoder.matches(loginRequest.getPassword(), user.getPassword());
+        if (!matches) {
+            return ResponseEntity.status(CommonConstant.BAD_REQUEST_CODE).body("Incorrect username or password");
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/service/impl/UserServiceImpl.java]: Use the predefined HTTP status code in the `package org.springframework.http`. for example: `HttpStatus.BAD_REQUEST`.

File: src/main/java/pro/quicksense/modules/util/EmailUtil.java [hunk 46]
```diff
@@ -0,0 +1,74 @@
...
+import org.springframework.mail.javamail.JavaMailSender;
+import org.springframework.stereotype.Service;
+import pro.quicksense.modules.common.CommonConstant;
+import pro.quicksense.modules.service.EmailService;
+
+import java.util.concurrent.TimeUnit;
+
+@Service
+@Slf4j
+public class EmailUtil implements EmailService {
```
Comment (cr7258) [inline src/main/java/pro/quicksense/modules/util/EmailUtil.java]:
```suggestion
public class EmailServiceImpl implements EmailService {
```
rename `EmailUtil` to `EmailServiceImpl`, and move the file under `impl` folder to make it consistent with `UserServiceImpl`.

**Comment by cr7258** [comment]
🐉 LGTM

**Pull request merged**